std/array.d

```diff
@@ -3519,6 +3519,74 @@ if (isDynamicArray!A)
         return cast(typeof(return))(_data ? _data.arr : null);
     }
 
+    /**
+     * Returns: The current length of the array
+     */
+    @property size_t length() const
+    {
+        return _data ? _data.arr.length : 0U;
+    }
+
+    /**
+     * Set a new length for the array
+     * Params:
+     *    a = The new length
+     */
+    @property void length(size_t newLength) @trusted
+    {
+        import core.stdc.string : memset, memcpy;
+        static void doInitialize(void *start, void *end, const void[] initializer)
+        {
+            if (initializer.length == 1)
+            {
+                memset(start, *(cast(ubyte*)initializer.ptr), end - start);
+            }
+            else
+            {
+                auto q = initializer.ptr;
+                immutable initsize = initializer.length;
+                for (; start < end; start += initsize)
+                    memcpy(start, q, initsize);
+            }
+        }
+
+        static if (!__traits(isZeroInit, T))
+            static immutable initializer = T.init;
+
+        if (_data)
+        {
+            if (newLength > _data.arr.length)
+            {
+                immutable oldLength = _data.arr.length;
+                reserve(newLength - oldLength);
+                _data.arr = _data.arr.ptr[0 .. newLength];
+                static if (__traits(isZeroInit, T))
+                    memset(_data.arr.ptr + oldLength, 0, (newLength - oldLength) * T.sizeof);
+                else
+                    doInitialize(
+                        _data.arr.ptr + oldLength,
+                        _data.arr.ptr + newLength,
+                        (cast(void*)&initializer)[0 .. T.sizeof]
+                    );
+                return;
+            }
+            _data.arr = _data.arr.ptr[0 .. newLength];
+            return;
+        }
+
+        reserve(newLength);
+        _data.arr = _data.arr.ptr[0 .. newLength];
+
+        static if (__traits(isZeroInit, T))
+            memset(_data.arr.ptr, 0, newLength * T.sizeof);
+        else
+            doInitialize(
+                _data.arr.ptr,
+                _data.arr.ptr + newLength,
+                (cast(void*)&initializer)[0 .. T.sizeof]
+            );
```

Comment:
```suggestion
            import core.lifetime : emplace;
            foreach (ref e; _data.arr.ptr[0 .. newLength])
                    emplace(&e);
```

Also, even though that obviates `doInitialize`, there's a `__traits(initSymbol, T)` for structs, classes and unions, introduced specifically to not have to have a `static immutable initiailzer = T.init;`.

Reply:
No. `__traits(initSymbol, T)` mimics the `TypeInfo.initializer`, not the `T.init`. For classes, it returns the emplaced memory block when dereferenced, not the actual reference.

Also `emplace` doesn't work here as arrays of abstract classes or structs with disabled constructors would not work. Although, I think `emplaceInitializer` can fit.